src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,196 @@
+        SOFT_MAX_ARRAY_LENGTH = Integer.MAX_VALUE - 8;
+        DEFAULT_CAPACITY_EMPTY_ELEMENT_DATA = new Object[]{};
     }
 
-    @Override
-    public void add(T value, int index) {
+    private Object[] elementData;
+    private int size;
 
+    public ArrayList() {
+        elementData = DEFAULT_CAPACITY_EMPTY_ELEMENT_DATA;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        Object[] objectsList = toArray(list);
+        int numNew = objectsList.length;
+        if (numNew == 0) {
+            return;
+        }
+        Object[] elementData;
+        final int size;
+        if (numNew > (elementData = this.elementData).length - (size = this.size)) {
+            elementData = grow(size + numNew);
+        }
+        System.arraycopy(objectsList, 0, elementData, size, numNew);
+        this.size = size + numNew;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, size);
+        return elementData(index);
     }
 
     @Override
     public void set(T value, int index) {
+        checkIndex(index, size);
+        elementData[index] = value;
+    }
 
+    @Override
+    public int size() {
+        return size;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return size == 0;
+    }
+
+    public Object[] toArray(List<T> list) {
+        Object[] objects = new Object[list.size()];
+        for (int i = 0; i < list.size(); i++) {
+            objects[i] = list.get(i);
+        }
+        return objects;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, size);
+        final Object[] elements = elementData;
+        @SuppressWarnings("unchecked")
+        T oldValue = (T) elements[index];
+        fastRemove(elements, index);
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int searchingIndex = foundElementIndex(element);
+        if (searchingIndex > NOT_EXISTING_INDEX) {
+            fastRemove(elementData, searchingIndex);
+            return element;
+        }
+        throw new NoSuchElementException("Not found such element.");
+    }
+
+    private int foundElementIndex(T element) {
+        final Object[] elements = elementData;
+        final int size = this.size;
+        int i = 0;
+        if (element == null) {
+            for (; i < size; i++) {
+                if (elements[i] == null) {
+                    return i;
+                }
+            }
+        } else {
+            for (; i < size; i++) {
+                if (element.equals(elements[i])) {
+                    return i;
+                }
+            }
+        }
+        return NOT_EXISTING_INDEX;
+    }
+
+    private void fastRemove(Object[] es, int i) {
+        final int newSize;
+        if ((newSize = size - 1) > i) {
+            System.arraycopy(es, i + 1, es, i, newSize - i);
+        }
+        es[size = newSize] = null;
     }
```

Comment:
This method is also redundant. Instead of this whole method you can just call something like this:
`System.arraycopy(elementData, index + 1, elementData, index, size - index - 1);`
also remember to decrement the 'size' after element removal